CodeU_Assignement_2/src/codeu_assignement_2/KemBinaryTree.java

```diff
@@ -0,0 +1,394 @@
+            return;
+        }
+        
+        /* Print all ancestors as long as didn't get to root */
+        while (curr != null)
+        {
+            System.out.print(curr.getData());
+            curr = curr.getParent();
+            if (curr != null)
+            {
+                System.out.print(",");
+            }
+        }
+        System.out.println(". Done printAncestors(..).");
+    }
+
+    /* Q2: Get all ancestors */
+    public KemTreeNode<E> getLowestCommonAncestors(KemTreeNode<E> child1, KemTreeNode<E> child2)
+    {
+        /* No common if at least one is null */
+        if ((child1 == null) || (child2 == null))
+        {
+            return null;
+        }
+        /* If one of them is the root - return it as the common */
+        if ((child1 == m_root) || (child2 == m_root))
+        {
+            return m_root;
+        }
+        /* Couts the edges between the node and the root */
+        int size_c1 = getSizeOfTraceChild2Root(child1);
+        int size_c2 = getSizeOfTraceChild2Root(child2);
+        if ((size_c1 == -1) || (size_c2 == -1))
+        {
+            return null;
+        }
+        /* Is root one of them */
+        if ((size_c1 == 0) || (size_c2 == 0))
+        {
+            return this.m_root;
+        }
+        
+        /* If larger than 0 */ 
+        KemTreeNode<E> common = m_root; /* If root not null, the worst case is root */
+        int min = (size_c2 > size_c1)? size_c1 : size_c2;
+        int c1_loc = size_c1;
+        int c2_loc = size_c2;
+        KemTreeNode<E> curr1 = child1; 
+        KemTreeNode<E> curr2 = child2;
+        while((c1_loc > 0) && (c2_loc > 0)) /* >, can skip checking root, and if 0 return root */
+        {
+            if (c1_loc > min) 
+            {
+                curr1 = curr1.getParent();
+                c1_loc--;
+            } 
+            else if (c2_loc > min) 
+            {
+                curr2 = curr2.getParent();
+                c2_loc--;
+            } 
+            else if (c1_loc == c1_loc) 
+            {
+                if (curr1 == curr2)
+                {
+                    common = curr2;
+                    break;
+                }
+                
+                curr1 = curr1.getParent();
+                curr2 = curr2.getParent();
+                c1_loc--;
+                c2_loc--;
+            }
+            else /* Something went wrong */
+            {
+                throw new IndexOutOfBoundsException();
+            }
+        }
+        
+        /* Return the most inner common */
+        return common;
+    }
+    
+    /* Create the trace from child to root, including the root and the child 
+        Return -1 if m_root is null
+                0 if it is the root that it isn't null
+               >0 for any other location on the path
+    
+    */
+    private int getSizeOfTraceChild2Root(KemTreeNode<E> child)
+    {
+        int ret_val = 0;
+        if (m_root == null) 

```

Comment:
I can throw assert instead? it is a corner-case that shouldn't really happen....